### PR TITLE
PMP : make it work with Epeck

### PR DESCRIPTION
**Pull request by janetournois**
This PR introduces CGAL::approximate_sqrt to avoid using to_double() and make the PMP package work with Epeck.

It also solves issue #326 

TODO: This PR is not ready for testing
@afabri and I should make it complete first

### CODE REVIEW

**Comment by lrineau** [comment]
The fairing seems the only user of `<CGAL/Polygon_mesh_processing/Weights.h>`. It should be documented that it is not exact.

**Comment by janetournois** [comment]
Skeletonization is too, in Mean_curvature_flow_skeletonization.h, but I don't know much about it
@sloriot should it be documented non-exact too?

**Comment by janetournois** [comment]
done for fairing

**Comment by lrineau** [comment]
That's right. Outside of PMP, there are two extra headers:
```
renoir ~/Git/cgal $ ack Weights.h ^build*/include -l
Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/fair_impl.h
Surface_mesh_deformation/include/CGAL/Surface_mesh_deformation.h
Surface_mesh_skeletonization/include/CGAL/Mean_curvature_flow_skeletonization.h
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/fair.h [hunk 1]
```diff
@@ -99,6 +99,8 @@ namespace internal {
 
   @pre `is_triangle_mesh(tmesh)`
 
+  @warning This function involves square root computations that will be done approximately.
+
```
Comment (lrineau) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/fair.h]: I do not think that is exact. That is not "square root computations".

I suggest:
> This function involves linear algebra, that is compute using a non-exact floating-point arithmetic.

or something similar.

**Comment by sloriot** [comment]
@lrineau I don't see your point here. What exactness would you expect from a function doing smoothing (fairing), deformation or skeleton extract that are by the way they are defined not exact? I don't see what you want to document.

**Comment by lrineau** [comment]
@afabri: This work was started two months ago. Do not forget this PR!

**Comment by sloriot** [comment]
@lrineau  what is blocking this PR?

**Comment by lrineau** [comment]
1. In the first comment, Jane said:
> This PR is not ready for testing

2. I made a suggestion, https://github.com/CGAL/cgal/pull/957#discussion-diff-57713443, that was left unanswered.

3. I am still not sure if the work of documenting which operations use inexact constructions is finished.

As far as we can see, Andreas and Jane simply stopped working on the PR. I am not the one responsible for the stalling.

**Comment by lrineau** [comment]
As for your last comment, https://github.com/CGAL/cgal/pull/957#issuecomment-202874539:
> @lrineau I don't see your point here. What exactness would you expect from a function doing smoothing (fairing), deformation or skeleton extract that are by the way they are defined not exact? I don't see what you want to document.

I agree that functions doing obviously inexact constructions may not be documented as such, the same way we do not document that Mesh_2 use inexact constructions for the corning looping. But even for those one, adding one sentence will not perturb the reader of the documentation.

**Comment by janetournois** [comment]
Sorry for leaving this PR behind. I will try to take care of it asap

**Comment by afabri** [comment]
No need to blame somebody. It's  just one PR among others.

**Comment by sloriot** [comment]
Locally building the polyhedron demo I get this error

       Algebraic_foundations/include/CGAL/number_utils.h:110:18: error: ‘to_double’ is not a member of ‘CGAL’
       return NT(sqrt(CGAL::to_double(nt)));`
                      ^

**Comment by janetournois** [comment]
I can't reproduce this error...

**Comment by janetournois** [comment]
is it better @sloriot ? It still compiles on msvc

File: Algebraic_foundations/include/CGAL/number_utils.h [hunk 2]
```diff
@@ -107,7 +107,7 @@ sqrt( const AS& x ) {
 template <typename NT>
 NT approximate_sqrt(const NT& nt, CGAL::Field_tag)
 {
-  return NT(sqrt(CGAL::to_double(nt)));
+  return NT(sqrt(to_double(nt)));
```
Comment (lrineau) [inline Algebraic_foundations/include/CGAL/number_utils.h]: If `NT` is `double`, then the call `to_double(nt)` will not compile.

The solution is to qualify the call: `CGAL::to_double`, but this definition of `approximate_sqrt` should be move to after the line 320, where `CGAL::to_double` is declared and defined.

Comment (janetournois) [inline Algebraic_foundations/include/CGAL/number_utils.h]: ah! ok!
thanks, I'm doing it

Comment (janetournois) [inline Algebraic_foundations/include/CGAL/number_utils.h]: done, see 	7dec6cd

**Comment by janetournois** [comment]
fixed

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/compute_normal.h [hunk 3]
```diff
@@ -40,19 +40,38 @@ namespace Polygon_mesh_processing{
...
   {
-    return CGAL::cross_product(p2 - p1, p0 - p1);
+    double norm = CGAL::approximate_sqrt(
+        traits.compute_squared_length_3_object()(v));
+    v = traits.construct_divided_vector_3_object()(v, norm);
+  }
+
+  template<typename Point
+         , typename GT = typename CGAL::Kernel_traits<Point>::Kernel>
+  typename GT::Vector_3
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/compute_normal.h]: Default template parameters are only allowed in c++11

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h [hunk 4]
```diff
@@ -284,19 +299,24 @@ namespace Polygon_mesh_processing {
...
   }
 
-  template<typename FaceRange, typename TriangleMesh>
-  double area(FaceRange face_range, const TriangleMesh& tmesh)
+  template<typename FaceRange, typename TriangleMesh > 
+  /* AF   what is wrong here???
+   typename CGAL::Kernel_traits<typename property_map_value<TriangleMesh,
+                                                            CGAL::vertex_point_t>::type>::Kernel::FT
+  */
+     double
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h]: incorrect indentation, Is the comment useless now ?

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/compute_normal.h (hunk 3 above)
Comment (lrineau) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/compute_normal.h]: ... for function templates.

For class templates that is allowed in C++98 as well.

File: Surface_mesh_skeletonization/include/CGAL/Mean_curvature_flow_skeletonization.h [hunk 5]
```diff
@@ -1320,12 +1322,22 @@ std::size_t collapse_short_edges();
   /// Compute an approximate vertex normal for all vertices.
   void compute_vertex_normal()
   {
+    namespace PMP = CGAL::Polygon_mesh_processing;
+
+    boost::unordered_map<face_descriptor, Vector> normals;
+    boost::associative_property_map<
+      boost::unordered_map<face_descriptor, Vector> > normals_pmap(normals);
+    PMP::compute_face_normals(m_tmesh, normals_pmap);
+
```
Comment (sloriot) [inline Surface_mesh_skeletonization/include/CGAL/Mean_curvature_flow_skeletonization.h]: IIRC, the polyhedron type used has ids in the faces, using a `std::vector` for storing the face normal would be more efficient. The vertex normal are recomputed at each contraction step so I guess this should make a difference.

**Comment by janetournois** [comment]
There are errors in the testsuite. The tests need more work and we figured out Epeck has issues.

**Comment by sloriot** [comment]
Already fixed, see my last commit

**Comment by janetournois** [comment]
it's not enough, see issue #1118 and errors in the `measures_test` in [this testsuite](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.9-Ic-59/Polygon_mesh_processing/TestReport_jtournois_x64_Cygwin-Windows10_MSVC2015-Debug-64bits.gz)

**Comment by janetournois** [comment]
and there is a `double` hard-coded in `measures.h`, that should be removed

**Comment by sloriot** [comment]
Errors [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.9-Ic-61/Polygon_mesh_processing/TestReport_jtournois_x64_Cygwin-Windows10_MSVC2015-Debug-64bits.gz), [there](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.9-Ic-61/Polygon_mesh_processing/TestReport_lrineau_CentOS5.gz) and [there](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.9-Ic-61/Polygon_mesh_processing/TestReport_lrineau_CentOS6-CXX11-Boost157.gz)

**Comment by janetournois** [comment]
this PR should be ready for testing.
Commit 0693d46 comments the issue spotted in issue #1118 . I plan to find temporary fixes if needed

**Comment by janetournois** [comment]
my last commit introduces the temporary fix I mentioned above.

It's temporary until Epeck is made stack-overflow free

I'm wondering what we should add to the doc to warn the user about the non-exactness of the functions that use the `operator+=`

**Comment by janetournois** [comment]
testing measures for Epeck is back! Ready for testing

**Comment by sloriot** [comment]
There is the following warning [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.9-Ic-68/Polygon_mesh_processing_Examples/TestReport_lrineau_CentOS5.gz):
`/mnt/testsuite/include/CGAL/Polygon_mesh_processing/internal/named_function_params.h:56: warning: direct base ‘CGAL::dummy_named_params’ inaccessible in ‘CGAL::pmp_bgl_named_params<CGAL::Epick, CGAL::geom_traits_t, CGAL::pmp_bgl_named_params<CGAL::Surface_mesh<CGAL::Point_3<CGAL::Epick> >::Property_map<CGAL::SM_Vertex_index, CGAL::Point_3<CGAL::Epick> >, boost::vertex_point_t, boost::no_property> >’ due to ambiguity`

btw I though `dummy_named_params` was renamed to `named_params_base`?

**Comment by janetournois** [comment]
you are absolutely right, I will fix that

**Comment by janetournois** [comment]
warnings and errors of testsuite CGAL-4.9-Ic-68 should be fixed now

**Comment by sloriot** [comment]
PMP testsuite is fully red

**Comment by janetournois** [comment]
I can't reproduce the compilation error on msvc2013, nor on msvc2015... Is it possible that this branch is not guilty for the red line?

**Comment by sloriot** [comment]
Master is more recent that your branch, and some code was added in `triangulate_faces` to handle the case of quads. In this code `double` is hardcoded.

**Comment by janetournois** [comment]
Oh ok! I will rebase the branch then. Thanks

**Comment by janetournois** [comment]
I rebased the branch and fixed the compilation error

**Comment by sloriot** [comment]
There is one warning remaining on g++ 4.1 but it is a bug of the compiler.

**Pull request merged**